### PR TITLE
Cleanup and refactoring for offline mode.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Prerequisite for showing data from previous app connection on disconnect.

Fixes https://github.com/flutter/devtools/issues/5101.

RELEASE_NOTE_EXCEPTION=[not user facing]

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/profiler/profiler_screen.dart [hunk 1]
```diff
@@ -88,6 +85,8 @@ class _ProfilerScreenBodyState extends State<ProfilerScreenBody>
 
     cancelListeners();
 
+    addAutoDisposeListener(controller.loadingOfflineData);
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/profiler/profiler_screen.dart:88]: Why are we registering a listener with no handler?

File: packages/devtools_app/lib/src/screens/profiler/profiler_screen_controller.dart [hunk 2]
```diff
@@ -110,13 +135,23 @@ class ProfilerScreenController extends DisposableController
...
-  String exportData() {
-    final encodedData =
-        _exportController.encode(ProfilerScreen.id, cpuProfileData!.toJson);
-    return _exportController.downloadFile(encodedData);
+  @override
+  OfflineScreenData screenDataForExport() =>
+      OfflineScreenData(ProfilerScreen.id, cpuProfileData!.toJson);
+
+  @override
+  FutureOr<void> processOfflineData(CpuProfileData offlineData) async {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/profiler/profiler_screen_controller.dart:143]: moved from profiler screen state

File: packages/devtools_app/lib/src/shared/offline_mode.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import 'connected_app.dart';
+import 'globals.dart';
+import 'primitives/auto_dispose.dart';
+
+class OfflineModeController {
+  bool get isOffline => _offlineMode.value;
+
+  ValueListenable<bool> get offlineMode => _offlineMode;
+
+  final _offlineMode = ValueNotifier(false);
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/offline_mode.dart:19]: Nit: `ValueNotifier<bool>(false)`

File: packages/devtools_app/lib/src/shared/offline_mode.dart [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  Map<String, dynamic> offlineDataJson = {};
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/offline_mode.dart:21]: Ubernit: `final offlineDataJson = <String, dynamic>{}`

File: packages/devtools_app/lib/src/shared/offline_mode.dart [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+}
+
+mixin OfflineScreenControllerMixin<T> on AutoDisposeControllerMixin {
+  final _exportController = ExportController();
+
+  ValueListenable<bool> get loadingOfflineData => _loadingOfflineData;
+  final _loadingOfflineData = ValueNotifier<bool>(false);
+
+  OfflineScreenData screenDataForExport();
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/offline_mode.dart:52]: Documentation?

File: packages/devtools_app/lib/src/shared/offline_mode.dart [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+  /// Exports the current screen data to a .json file.
+  void exportData() {
+    final encodedData = _exportController.encode(screenDataForExport().json);
+    _exportController.downloadFile(encodedData);
+  }
+}
+
+class OfflineScreenData {
+  OfflineScreenData(this.screenId, this.data);
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/offline_mode.dart:70]: Consider making these required named parameters for readability.

**Review by bkonyi — APPROVED** [review]

File: packages/devtools_app/lib/src/screens/profiler/profiler_screen.dart (hunk 1 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/screens/profiler/profiler_screen.dart:88]: by default the handler is a call to setState

**Pull request merged**